<a href="https://colab.research.google.com/github/VictorHugoKlein/Boston_Housing_for_model_selection/blob/main/Case_Boston_Housing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import numpy as np
from sklearn.svm import SVR
from xgboost import XGBRegressor

In [ ]:
df = pd.read_csv("/content/Boston.csv", sep =",")

In [ ]:
df.head()

Unnamed: 0     crim    zn  indus  chas    nox     rm   age     dis  rad  \
0           1  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1   
1           2  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2   
2           3  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2   
3           4  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3   
4           5  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3   

   tax  ptratio   black  lstat  medv  
0  296     15.3  396.90   4.98  24.0  
1  242     17.8  396.90   9.14  21.6  
2  242     17.8  392.83   4.03  34.7  
3  222     18.7  394.63   2.94  33.4  
4  222     18.7  396.90   5.33  36.2

In [ ]:
df.drop(["Unnamed: 0"], axis=1, inplace=True)

In [ ]:
df.head()

crim    zn  indus  chas    nox     rm   age     dis  rad  tax  ptratio  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222     18.7   

    black  lstat  medv  
0  396.90   4.98  24.0  
1  396.90   9.14  21.6  
2  392.83   4.03  34.7  
3  394.63   2.94  33.4  
4  396.90   5.33  36.2

In [ ]:
df.columns

Index(['crim', 'zn', 'indus', 'chas', 'nox', 'rm', 'age', 'dis', 'rad', 'tax',
       'ptratio', 'black', 'lstat', 'medv'],
      dtype='object')

In [ ]:
X = df[['crim', 'zn', 'indus', 'chas', 'nox', 'rm', 'age', 'dis', 'rad', 'tax', 'ptratio', 'black', 'lstat']]

Y = df[['medv']]

#MODELING TECHNIQUES

1.Regressão Linear do SKLEARN

  https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html

2.Support Vector Regression do SKLEARN

  https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVR.html

3.Decision Tree Regression do XGBoost

https://xgboost.readthedocs.io/en/latest/python/python_api.html

#TEST DESIGN

##Dataset split:

Separação de Train/Test dataset padrão com 20% de massa para teste via método SKLEARN

##Métrica de avaliação do modelo:

Validação da métrica MSE e RMSE para penalizar grandes erros de previsão. Utilizando o método do SKLEARN.

<https://scikit-learn.org/stable/modules/generated/sklearn.metrics.mean_squared_error.html>


In [ ]:
x_train, x_test, y_train, y_test = train_test_split(X,Y, test_size=0.2, random_state=42)

#TECNICA 1: REGRESSÃO LINEAR

In [ ]:
regLinear = LinearRegression().fit(X_train,Y_train)

In [ ]:
##Predição do valor Y através do modelo de regressão linear (já treinado)

yLinear = regLinear.predict(x_test)

In [ ]:
#Encontrando o mínimo erro quadrado entre os dados existentes que foram usados
#para teste (y_test) e o resultado da predição (yLinear)

MSELinear = mean_squared_error(y_test, yLinear)

In [ ]:
print("MSE Linear:",MSELinear)

MSE Linear: 24.291119474973478


In [ ]:
print("RSEM Linear:",np.sqrt(MSELinear))

RSEM Linear: 4.928602182665332


#TECNICA 2: SVR - Support Vector Regression

In [ ]:
regSVR = SVR().fit(x_train, y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [ ]:
ySVR = regSVR.predict(x_test)

In [ ]:
print(ySVR)

[22.5061509  24.44637611 15.48342923 23.47861138 15.85575665 20.41359379
 22.22729603 19.30750829 16.00684204 20.69013802 22.5948442  22.17932651
 13.22283187 20.88634606 22.74343046 16.11962825 23.6578174  15.50113391
 24.80649803 15.86119502 24.85903189 24.42774747 21.65632951 23.06890895
 15.18170664 15.47518332 21.28809591 13.27760452 22.58811671 20.84375287
 22.39598564 23.17949952 15.87553566 15.81646768 15.26751873 18.53736675
 23.75510385 24.20893309 23.03137207 23.35824097 20.74014355 24.26097498
 25.29050354 22.45914878 23.21111935 15.87406945 21.6649305  23.42123755
 16.08170391 23.06164042 22.35542867 24.67123285 22.08392839 21.02076269
 23.7698124  16.48115196 15.85681759 24.99519292 23.87040225 22.19275612
 23.61440768 25.38561307 23.18766117 21.82639671 21.46613519 22.43757636
 15.82206507 23.86467246 24.9310726  13.3354447  24.23171731 16.22743397
 14.35580648 23.47121935 21.03977262 15.66672407 20.63701249 25.28324943
 15.81293928 15.0221233  22.55700153 15.01570683 23

In [ ]:
MSESVR = mean_squared_error(y_test, ySVR)

In [ ]:
print("MSESRV:",MSESVR)
print("RMSE SVR:", np.sqrt(MSESVR))

MSESRV: 52.8383657679667
RMSE SVR: 7.269000327965785


#TECNICA 3.DECISION TREE REGRESSION (XGBOOST)

In [ ]:
regXGB = XGBRegressor().fit(x_train, y_train)

In [ ]:
yXGB = regXGB.predict(x_test)

In [ ]:
MSEXGB = mean_squared_error(y_test, yXGB)

In [ ]:
print("MSE XGB:",MSEXGB)
print("RMSE XGB:", np.sqrt(MSEXGB))

MSE XGB: 6.909231565384943
RMSE XGB: 2.62854171840299


#ESCOLHA DO MODELO:

Concluímos que o melhor modelo é o Decision Tree do XGBoost, já que esse possui um RMSE menor que dos demais modelos. Ou seja, o Decision Tree apresenta menores desvios dos erros.

#Utilização do GridSearchCV()

Com o modelo selecionado, o próximo passo é aprimorá-lo a partir da utilização da função GridSearchCV e alguns parâmetros.

Essa função é um otimizador de hiperparâmetros.

https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html#sklearn.model_selection.GridSearchCV


In [ ]:
from sklearn.model_selection import GridSearchCV

In [ ]:
regXGB.get_params().keys()

dict_keys(['objective', 'base_score', 'booster', 'callbacks', 'colsample_bylevel', 'colsample_bynode', 'colsample_bytree', 'device', 'early_stopping_rounds', 'enable_categorical', 'eval_metric', 'feature_types', 'gamma', 'grow_policy', 'importance_type', 'interaction_constraints', 'learning_rate', 'max_bin', 'max_cat_threshold', 'max_cat_to_onehot', 'max_delta_step', 'max_depth', 'max_leaves', 'min_child_weight', 'missing', 'monotone_constraints', 'multi_strategy', 'n_estimators', 'n_jobs', 'num_parallel_tree', 'random_state', 'reg_alpha', 'reg_lambda', 'sampling_method', 'scale_pos_weight', 'subsample', 'tree_method', 'validate_parameters', 'verbosity'])

In [ ]:
parameters = {
    "max_depth": [5, 6, 7],
    "learning_rate":[0.1, 0.2, 0.3],
    "objective": ['reg:squarederror'],
    "booster": ['gbtree'],
    "n_jobs": [5],
    "gamma": [0,1],
    "min_child_weight": [1,3],
    "max_delta_step": [0,1],
    "subsample": [0.5,1]
}

In [ ]:
xgbGrid = GridSearchCV(XGBRegressor(), parameters, refit='neg_mean_squared_error', verbose=True)

In [ ]:
xgbGridModel = xgbGrid.fit(x_train, y_train)

Fitting 5 folds for each of 144 candidates, totalling 720 fits


In [ ]:
xgbGridModel.best_params_

{'booster': 'gbtree',
 'gamma': 0,
 'learning_rate': 0.2,
 'max_delta_step': 0,
 'max_depth': 5,
 'min_child_weight': 1,
 'n_jobs': 5,
 'objective': 'reg:squarederror',
 'subsample': 0.5}

In [ ]:
yGrid = xgbGridModel.predict(x_test)

In [ ]:
MSEGrid = mean_squared_error(y_test, yGrid)

In [ ]:
print("MSE XGB Grid:",MSEGrid)
print("RMSE XGB Grid:", np.sqrt(MSEGrid))

MSE XGB Grid: 8.2975112347883
RMSE XGB Grid: 2.8805400942858443


#CONCLUSÃO
O modelo default funcionou melhor que o modelo parametrizado, já que o primeiro apresentou um RMSE de 2.63 e o segundo um RMSE de 2.88.